src/event/cloudevent.ts

```diff
@@ -0,0 +1,122 @@
+const { v4: uuidv4 } = require("uuid");
+
+import { CloudEventV1, validateV1, CloudEventV1Attributes } from "./v1";
+import { CloudEventV03, validateV03, CloudEventV03Attributes } from "./v03";
+import { ValidationError } from "./validation";
+
+export const enum Version { V1 = "1.0", V03 = "0.3" };
+
+
+/**
+ * A CloudEvent describes event data in common formats to provide
+ * interoperability across services, platforms and systems.
+ * @see https://github.com/cloudevents/spec/blob/v1.0/spec.md
+ */
+export class CloudEvent implements CloudEventV1, CloudEventV03 {
+  id: string;
+  type: string;
+  source: string;
+  specversion: Version;
+  datacontenttype?: string;
+  dataschema?: string; // TODO: Is a URI type needed?
+  subject?: string;
+  time?: string|Date;
+  data?: any;
+  data_base64?: any;
+  [key: string]: any; // Extensions should not exist as it's own object, but instead be keyed directly from the event
+
+  // V03 deprecated attributes
+  schemaurl?: string;
+  datacontentencoding?: string;
+
+  constructor(event: CloudEventV1 | CloudEventV1Attributes | CloudEventV03 | CloudEventV03Attributes) {
+    // copy the incoming event so that we can delete properties as we go
+    // everything left after we have deleted know properties becomes an extension
```

Comment:
what is the reasoning for not keeping extensions its own object within the event?

Reply:
Because [in the spec](https://github.com/cloudevents/spec/blob/master/primer.md#json-extensions), when a CloudEvent is serialized to for wire transport, the extensions are on the main event itself and not in an `extensions` object.

> As mentioned in the Attributes section of the JSON Event Format for CloudEvents specification, CloudEvent extension attributes are serialized as siblings to the specification defined attributes - meaning, at the top-level of the JSON object. 

With the previous implementation, there was a `CloudEvent.format()` function that would serialize things properly. But one of the goals of this big rewrite was to make it so that a `CloudEvent` could be serialized for transport with little more than `JSON.stringify(event)`, and that a new event could be created with the extensions inline, e.g. `new CloudEvent({ extension: 'some value'});`.

Reply:
sounds good.  I do remember seeing an article why using `delete` might not be as efficient, but of course i can't find it,  and it was probably more than 5 yrs old,  so it is possible that info could be outdated